### PR TITLE
Fix 'quick info' and 'use explicit type' for implicit lambdas

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 1]
```diff
@@ -261,7 +261,7 @@ public static bool IsNormalAnonymousType([NotNullWhen(returnValue: true)] this I
             => symbol.IsAnonymousType() && !symbol.IsDelegateType();
 
         public static bool IsAnonymousDelegateType([NotNullWhen(returnValue: true)] this ISymbol? symbol)
-            => symbol.IsAnonymousType() && symbol.IsDelegateType();
+            => symbol.IsDelegateType() && (symbol.IsAnonymousType() || !symbol.CanBeReferencedByName);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: vb and C# are a bit weird here.   VB represents anonymous lambdas as actual .IsAnonymousType.  C# does not, but instead has CanBeReferencedByName=false.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 2]
```diff
@@ -343,12 +342,18 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
...
+            var isAnoymousDelegate = symbol.TypeKind == TypeKind.Delegate && !symbol.CanBeReferencedByName;
             if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.UseErrorTypeSymbolName) &&
                 partKind == SymbolDisplayPartKind.ErrorTypeName &&
                 string.IsNullOrEmpty(symbolName))
             {
                 builder.Add(CreatePart(partKind, symbol, "?"));
             }
+            else if (isAnoymousDelegate)
+            {
+                var name = $"<anonymous delegate>";
```
Comment (jmarolf) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:354]: Does this need to be localized?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (jmarolf) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: Was there an explicit decision here by the compiler team to use `CanBeReferencedByName` over `IsAnonymousType` as the api shape? Wish these could be consistent

**Review by jmarolf — CHANGES_REQUESTED** [review]
:clock1:

**Review by jmarolf — APPROVED** [review]
:shipit:

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:354]: no.  we do the same here: https://github.com/dotnet/roslyn/blob/5bce330dc6294f059e6cf69a2ee77b02e141c804/src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs#L443-L447

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: @cston any thoughts on why VB/C# differ here wrt anonymous delegates?

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:354]: >$ [](http://example.com/codeflow?start=27&length=1)

Minor: `$` is not needed.

And perhaps just inline the value in the call below.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:354]: sure.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (cston) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: It feels like `SynthesizedDelegateSymbol.IsAnonymousType` should return `true`. Please feel free to change that or log an issue.

**Review by cston — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jinujoseph Need approval on this.  This is for some high profile ugliness cases we get with the new anonymous-lambda work that compiler just merged in.

**Comment by cston** [comment]
@dotnet/roslyn-compiler for a second review, thanks.

**Comment by CyrusNajmabadi** [comment]
Moving over to main.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 3]
```diff
@@ -338,17 +337,23 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
 
             var partKind = GetPartKind(symbol);
 
-            if (symbolName == null)
-            {
-                symbolName = symbol.Name;
-            }
+            symbolName ??= symbol.Name;
+
+            var renderSimpleAnoymousDelegate =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:342]: >Anoymous [](http://example.com/codeflow?start=28&length=8)

Typo.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 4]
```diff
@@ -338,17 +337,23 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
...
 
             if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.UseErrorTypeSymbolName) &&
                 partKind == SymbolDisplayPartKind.ErrorTypeName &&
                 string.IsNullOrEmpty(symbolName))
             {
                 builder.Add(CreatePart(partKind, symbol, "?"));
             }
+            else if (renderSimpleAnonymousDelegate)
+            {
+                builder.Add(new SymbolDisplayPart(SymbolDisplayPartKind.DelegateName, null, "<anonymous delegate>"));
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:355]: Are these things that we try to localize? I'm not familiar with our policy on that for this type.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (333fred) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: I'd agree with Chuck.

**Review by 333fred — APPROVED** [review]
LGTM, assuming my question around localization is answered with "no, we don't do that for this type."

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 1 above)
Comment (cston) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:264]: [https://github.com/dotnet/roslyn/issues/55292](https://github.com/dotnet/roslyn/issues/55292)

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:355]: we don't do it elsewhere in this type for things like anonymous-types.  so i'm keeping with that pattern.

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs [hunk 5]
```diff
@@ -7597,5 +7597,36 @@ struct B
...
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        $$var v = (int i) => i.ToString();
+    }
+}",
+                MainDescription("delegate TResult System.Func<in T, out TResult>(T arg)"),
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:7612]: Shouldn't this just be `System.Func<...>`? I don't understand the combination of `System.Func...` with `delegate` and a method signature.

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs [hunk 6]
```diff
@@ -7806,5 +7806,51 @@ class A {
...
+
+        [Fact]
+        public void TestSynthesizedAnonymousDelegateType1()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        var v = (ref int i) => i.ToString();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:7818]: Could you include a similar test for a "simple signature" (no `ref`)?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs [hunk 7]
```diff
@@ -319,17 +319,13 @@ public static bool ContainsAnonymousType([NotNullWhen(returnValue: true)] this I
 
         private static bool ContainsAnonymousType(INamedTypeSymbol type)
```
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: Can we remove this method and overload on line 309?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 6)

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:7612]: So Func is a delegate, so we show that (just like we say if something is a class/struct). And for delegates we show the signature as that's what's relevant a lot of the time.  So this is standard for how we've shown this information for about 15 years :-)

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:7818]: Sure!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs (hunk 7 above)
Comment (jcouv) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: Kindly ping on this question

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6) modulo question left about removing method rather than updating it.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: i don't really know what you mean.  How would we remove it?

Comment (jcouv) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: I'm suggesting we delete lines 309 through 332.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: Why would we do that?

Comment (jcouv) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: Arg. My bad. When I'd done a FindAllRefs, I'd only found those two methods referencing each other... But in actuality they are used...
Good to go then. Thanks

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: basically, i don't understand the motivation here.  why would we remove these helpers?  and what is your suggestion on what to do about the breakage?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:320]: got it :)  ok.

**Pull request merged**